homeassistant/components/rabbitair/fan.py

```diff
@@ -0,0 +1,148 @@
+"""Support for Rabbit Air fan entity."""
+from __future__ import annotations
+
+from typing import Any
+
+from rabbitair import Mode, Model, Speed
+
+from homeassistant.components.fan import FanEntity, FanEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.percentage import (
+    ordered_list_item_to_percentage,
+    percentage_to_ordered_list_item,
+)
+
+from .const import DOMAIN
+from .coordinator import RabbitAirDataUpdateCoordinator
+from .entity import RabbitAirBaseEntity
+
+SPEED_LIST = [
+    Speed.Silent,
+    Speed.Low,
+    Speed.Medium,
+    Speed.High,
+    Speed.Turbo,
+]
+
+PRESET_MODE_AUTO = "Auto"
+PRESET_MODE_MANUAL = "Manual"
+PRESET_MODE_POLLEN = "Pollen"
+
+PRESET_MODES = {
+    PRESET_MODE_AUTO: Mode.Auto,
+    PRESET_MODE_MANUAL: Mode.Manual,
+    PRESET_MODE_POLLEN: Mode.Pollen,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up a config entry."""
+    coordinator: RabbitAirDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities([RabbitAirFanEntity(coordinator, entry)])
+
+
+class RabbitAirFanEntity(RabbitAirBaseEntity, FanEntity):
+    """Fan control functions of the Rabbit Air air purifier."""
+
+    _attr_supported_features = FanEntityFeature.PRESET_MODE | FanEntityFeature.SET_SPEED
+
+    def __init__(
+        self,
+        coordinator: RabbitAirDataUpdateCoordinator,
+        entry: ConfigEntry,
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(coordinator, entry)
+
+        if self._is_model(Model.MinusA2):
+            self._attr_preset_modes = list(PRESET_MODES)
+        elif self._is_model(Model.A3):
+            # A3 does not support Pollen mode
+            self._attr_preset_modes = [
+                k for k in PRESET_MODES if k != PRESET_MODE_POLLEN
+            ]
+
+        self._attr_speed_count = len(SPEED_LIST)
+
+        self._get_state_from_coordinator_data()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._get_state_from_coordinator_data()
+        super()._handle_coordinator_update()
+
+    def _get_state_from_coordinator_data(self) -> None:
+        """Populate the entity fields with values from the coordinator data."""
+        data = self.coordinator.data
+
+        # Power on/off
+        self._attr_is_on = data.power
+
+        # Speed as a percentage
+        if data.speed is None:
+            self._attr_percentage = None
+        elif data.speed is Speed.SuperSilent:
+            self._attr_percentage = 0
```

Comment:
0 means the fan is off according to the [documentation](https://developers.home-assistant.io/docs/core/entity/fan). Maybe use 1 instead if this is a very low speed?

@bdraco WDYT?

Reply:
1 or whatever super silent is as a percentages of maximum speed makes sense to me

Reply:
Will use 1% for Super Silent mode.